add _structure.id child key data names  to GEOM*

rowlesmr opened the pull request:
will close #22 

`GEOM*` categories added with `_structure.id` as a key data name.

I don't know if the dREL needs to be altered or not.

Do there need to be any further data names added? I'm still not sure how categories and their children interact...

rowlesmr committed:
add _structure.id child key data names  to GEOM*

rowlesmr committed:
updated update dates

vaitkus commented on multi_block_core.dic:
```diff
@@ -668,6 +668,426 @@ save_space_group_wyckoff.space_group_id
+    _method.purpose               Evaluation
+    _method.expression
+;
+    dmin =  _geom.bond_distance_min
```
Line 920 uses `label` as  defined in line 903 (`label = m2.label`). Might be an optimisation of some sort.

Not sure about the answer to the remainder of the question.

jamesrhester replied:
Yes, I don't think there is a need to update the dREL snippets before this can be accepted. By not including the dREL in the revised category definition, it is effectively banished from the composite dictionary, as the new category definition replaces the old category definition. Updating the dREL can be a TODO.

The dREL for the individual data names should operate without alteration, as per [my other comment today](https://github.com/COMCIFS/MultiBlock_Dictionary/pull/24#issuecomment-2942566615).

rowlesmr replied:
OK. Leaving out dREL for this PR.

rowlesmr commented:
> I don't know if the dREL needs to be altered or not.

Looking through the dREL, they also seem to reference `_model_site.radius_bond`, `.radius_contact`, `.mole_index`, `.fract_xyz`, `.Cartn_xyz`, `.label`, and `.symop`. Does `MODEL` and `MODEL_SITE` need a `_structure.id`, too?

rowlesmr commented:
And for that matter, what about `VALENCE`, `VALENCE_PARAM`, and `VALENCE_REF`?

vaitkus commented:
> And for that matter, what about `VALENCE`, `VALENCE_PARAM`, and `VALENCE_REF`?

I think that `VALENCE`, `VALENCE_PARAM`, and `VALENCE_REF` do not need a structure id since, IMHO, they were not properly migrated from DDL1 to DDLm and should be reverted to their previous semantics (see issue https://github.com/COMCIFS/cif_core/issues/62).

rowlesmr commented:
> I don't know if the dREL needs to be altered or not.

I think it does.

Take `GEOM_BOND` for instance.

It is currently

```
    dmin =  _geom.bond_distance_min

    Loop  m1  as  model_site  :i  {

       rad   =   m1.radius_bond + _geom.bond_distance_incr

       Loop  m2  as  model_site  :j  {

          If (i==j or m1.mole_index != m2.mole_index) Next

          d =  Norm (m1.Cartn_xyz - m2.Cartn_xyz)

          If (d<dmin or d>(rad+m2.radius_bond)) Next

          geom_bond( .atom_site_label_1        =  m1.label,
                     .atom_site_label_2        =  m2.label,
                     .site_symmetry_1          =  m1.symop,
                     .site_symmetry_2          =  m2.symop,
                     .distance  =  d  )
     }   }
```

I believe it should be

``` 
    dmin =  _geom.bond_distance_min

    Loop  m1  as  model_site  :i  {

       rad   =   m1.radius_bond + _geom.bond_distance_incr

       Loop  m2  as  model_site  :j  {

          If (m1.structure_id != m2.structure_id or
              i==j or
              m1.mole_index != m2.mole_index)
              Next

          d =  Norm (m1.Cartn_xyz - m2.Cartn_xyz)

          If (d<dmin or d>(rad+m2.radius_bond)) Next

          geom_bond( .atom_site_label_1        =  m1.label,
                     .atom_site_label_2        =  m2.label,
                     .site_symmetry_1          =  m1.symop,
                     .site_symmetry_2          =  m2.symop,
                     .distance                 =  d,
                     .structure_id             =  m1.structure_id )
     }   }
```

the `If` in the middle might be the wrong place to put it, but it does at least only calc the distance if the `model_site`s come from the same `structure_id`. I think.

rowlesmr committed:
Add structure_id to MODEL

rowlesmr committed:
add structure_id to MODEL_SITE

rowlesmr committed:
Revert "add structure_id to MODEL_SITE"

This reverts commit 2cca46a0e0c4336bdaa0609e9c8e0044457823fe.

rowlesmr committed:
Revert "Add structure_id to MODEL"

This reverts commit ec5dd16b4ec4b00f04324bac5e408647b9055d0a.

rowlesmr committed:
add structure_id to MODEL

rowlesmr committed:
add structure_id to MODEL_SITE

rowlesmr committed:
update GEOM_ANGLE dREL

rowlesmr committed:
update GEOM_BOND dREL

rowlesmr committed:
update dREL GEOM_TORSION

rowlesmr committed:
refactor structure_id in dREL GEOM_ANGLE

rowlesmr committed:
refactor structure_id in dREL for GEOM_BOND

rowlesmr committed:
refactor structure_id in dREL for GEOM_TORSION

rowlesmr commented:
The dREL in `GEOM_CONTACT` and `MODEL_SITE` is difficult.

rowlesmr commented:
> I know the `_structure.id` of `m1`. How can I use that to get the respective `_space_group.id` of `m1`, given that `_structure.space_group_id` exists?

I think the answer to this lies in paragraph 3 of _dREL: A Relational Expression Language for Dictionary Methods_


> Data References in dREL. Syntactically, reference to a datavalue in dREL is achieved as follows.

>1. A data item is most commonly referenced by its internaldata name, constructed from a concatenation of the valueof the attribute name.category_id, <category>, and the value of the attribute name.object_id, <object>, as <category>.<object>.

>2. Reference to all of the multiple values of items in a Loop category is achieved by referring to the <category> component; this automatically provides the primary keysto every packet in a loop list.

>3. Where a value (fkey) in one Loop category A represents a foreign key (a data value that is the primary key) to a diﬀerent Loop category B, the construct <B>[A.fkey].<object> accesses the value <B>.<object> for the speciﬁc loop packet identiﬁed by A.fkey.

>4. For a data item deﬁned as type List, Array, or Matrix individual values are referenced by <category>.<object>[n,m,...] where n, m,... are the numerical indices to elements in that item.

>5. For a data item deﬁned as type Table an individual value is referenced by <category>.<object>[label] where label is the tag speciﬁc to the element.

rowlesmr commented:
> I know the `_structure.id` of `m1`. How can I use that to get the respective `_space_group.id` of `m1`, given that `_structure.space_group_id` exists?

I think the answer to this lies in paragraph 3 of _dREL: A Relational Expression Language for Dictionary Methods_



> Data References in dREL. Syntactically, reference to a datavalue in dREL is achieved as follows.
>1. A data item is most commonly referenced by its internaldata name, constructed from a concatenation of the valueof the attribute `name.category_id`, \<_category_\>, and the value of the attribute `name.object_id`, \<_object_\>, as \<_category_\>.\<_object_\>.
>2. Reference to all of the multiple values of items in a Loop category is achieved by referring to the \<category\> component; this automatically provides the primary keys to every packet in a loop list.
>3. Where a value (_fkey_) in one Loop category _A_ represents a foreign key (a data value that is the primary key) to a diﬀerent Loop category _B_, the construct \<_B_\>[_A.fkey_].\<_object_\> accesses the value \<_B_\>.\<_object_\> for the speciﬁc loop packet identiﬁed by _A.fkey_.
>4. For a data item deﬁned as type _List_, _Array_, or _Matrix_ individual values are referenced by \<_category_\>.\<_object_\>[_n,m,..._] where _n, m,.._. are the numerical indices to elements in that item.
>5. For a data item deﬁned as type Table an individual value is referenced by \<_category_\>.\<_object_\>[_label_] where label is the tag speciﬁc to the element.

jamesrhester commented:
> I think the answer to this lies in paragraph 3 of dREL: A Relational Expression Language for Dictionary Methods

We have slightly expanded the behaviour of dREL so that the equivalent of point 3 tries very hard to guess what the values of foreign keys should be when accessing other categories. See [here](https://github.com/COMCIFS/dREL/blob/main/drel_changes_2.rst). 

The text in Volume G at the moment reads:

> Whenever there is no ambiguity, row access subscriptions of the form `<category_other>[.name1=value1,.name2=values2,...]` may be simplified. In particular, if such a row access subscription appears in a method associated with data name `<category_home>.<object_id>`, those key data names of `<category_other>` for which a one-to-one relationship exists with key data names in `<category_home>` may be elided from row access expressions. The values of the elided attribute names are taken from the values of the corresponding attributes in `<category_home>` for the row that is currently selected. A data name is “related” to another data name if a common parent data name can be found via `_name.linked_item_id` links. 

> Note that this principle can only apply where there is a one-to-one mapping between the key data names in the two categories so that ambiguity is avoided. For example, the GEOM_BOND category has two key data names that are both children of the _atom_site.label key data name of the ATOM_SITE category, and therefore references to the ATOM_SITE category of the form atom_site.<object_id> in dREL methods associated with GEOM_BOND data names are ambiguous. Similarly, references to the GEOM_BOND category in dREL methods associated with data names in the ATOM_SITE category cannot elide key data names related to _atom_site.label, as the particular related data name in GEOM_BOND is ambiguous.

Here's a diagram that attempts to show this:
![drel-row-resolution](https://github.com/user-attachments/assets/2ebf0661-f8d8-421d-a407-5389986c038d)

What this means for the present discussion is that, for non-category methods, the value of `_structure.id` that is relevant for any out-of-category accesses is just the same value that is set on the current row, so doesn't need to be explicit. This has been designed to minimise disruption when we add new category keys.

rowlesmr commented:
> Here's a diagram that attempts to show this:

That's a 404 error on the image for me.

jamesrhester commented:
> > Here's a diagram that attempts to show this:
> 
> That's a 404 error on the image for me.

It looks like a normal Github link to me, not sure why that's happening.

rowlesmr commented:
> > > Here's a diagram that attempts to show this:
> > 
> > 
> > That's a 404 error on the image for me.
> 
> It looks like a normal Github link to me, not sure why that's happening.

Maybe a permissions thing?

rowlesmr commented on multi_block_core.dic:
```diff
@@ -668,6 +668,603 @@ save_space_group_wyckoff.space_group_id
+    _type.contents                Word
+
+save_
+
```
As I'm on shaky ground when dealing with categories and subcategories and their keys:

So no `_geom_*.structure_id`s at all. I'm cool with that.

There is `_geom.model_id` as a key; are there `_geom_*.model_id`s as keys? or are there `_geom_*.geom_model_id`s as keys?

There is `_model.id` as a key and `_model.structure_id` as a non-key data name. Is there then `_model_site.model_id` as a key? `MODEL_SITE` then doesn't need a structure_id, as it has a model.id, and can find structure_id from that....

rowlesmr replied:
I'll have a go tonight

jamesrhester replied:
Yes, this will have to be updated if we are happy with this change.

rowlesmr commented:
> So I think the solution is to enclose most of the whole original method with a loop over `_structure.id`, and then to explicitly use the current loop value of `_structure.id` in category accesses.

I'll give that a bash

rowlesmr committed:
Revert "refactor structure_id in dREL for GEOM_TORSION"

This reverts commit f74abe5429d626533582fda350d969b7395214b4.

rowlesmr committed:
Revert "refactor structure_id in dREL for GEOM_BOND"

This reverts commit 9776ed3b8884030f0f863948bb8b8c91ab6ebd41.

rowlesmr committed:
Revert "refactor structure_id in dREL GEOM_ANGLE"

This reverts commit 8002d06461c2d4279d0d7cfa0c3b76d1cd6e9266.

rowlesmr committed:
Revert "update dREL GEOM_TORSION"

This reverts commit 8bd16b6329822a1f571f86ead73d357d26481ad0.

rowlesmr committed:
Revert "update GEOM_BOND dREL"

This reverts commit 9db723c19ef3e5ea0427f36408891f1a42c7cd8d.

rowlesmr committed:
Revert "update GEOM_ANGLE dREL"

This reverts commit 91163c1c1a7b0a309e119d02ad14039199156c32.

rowlesmr committed:
add _model.id to a bunch of stuff

MODEL as _model.id as key and _model.structure_id as a normal dataname
GEOM* has _geom*.model_id as key
MODEL_SITE has _model_site.model_id as key

dREL still needs updating

rowlesmr committed:
fix some typos, update changelist

rowlesmr committed:
remove dREL

from MODEL_SITE and GEOM*

rowlesmr commented:
I've added `_model.id` and `_model.structure_id`, and `_geom_*.model_id` and `_model_site.model_id`.

I've removed the dREL to allow this change to go through with the view to adding it back later.

What say you?

jamesrhester committed:
Improve wording for _model.id linked data names
